{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2de96290-d64c-4c82-bd6d-fe69fcc1dad3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "from katana import remote\n",
    "from katana.remote import export_data, import_data\n",
    "\n",
    "client = remote.Client()\n",
    "client.server_version\n",
    "\n",
    "print(\"--\")\n",
    "\n",
    "\n",
    "#  Nothing going on here- left unchanged\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e89769fd-629e-419f-bb3f-55a0d60b0b73",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from config import hyperparams\n",
    "\n",
    "input_config = hyperparams.load_input_config()\n",
    "input_config\n",
    "\n",
    "\n",
    "#  Nothing going on here- left unchanged\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "885f5f2f-d988-4f60-bb56-9d430efd77e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from timeit import default_timer\n",
    "\n",
    "# Import the module that uses Dask to import drug data\n",
    "from src import dask_ingestion\n",
    "\n",
    "graph = client.create_graph(num_partitions=input_config.num_partitions)\n",
    "\n",
    "\n",
    "#  if input_config.use_train_rdg:\n",
    "#     print(f\"Import pretrained graph from: {input_config.trained_rdg_path}\")\n",
    "#     import_data.rdg(graph, input_config.trained_rdg_path)\n",
    "#  else:\n",
    "#     print(\"Generate the graph with data from source\")\n",
    "#     dask_ingestion.generate_deepcdr_graph(graph, input_config)\n",
    "\n",
    "import_data.rdg(graph, input_config.trained_rdg_path)\n",
    "#  dask_ingestion.generate_deepcdr_graph(graph, input_config)\n",
    "    \n",
    "print(\"--\")\n",
    "\n",
    "\n",
    "#  RDG import works, CSV fails\n",
    "#\n",
    "\n",
    "#  Can we start naming the created graphs ?  In my shared server, there are a ton of orphan graphs.\n",
    "#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46bb76e5-bfef-4333-a343-913e303ecb3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "##################################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d99d580-f5b1-471b-b540-d24fab04a24c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from src import katana_pipeline\n",
    "\n",
    "rec_pipeline = katana_pipeline.RecipePipeline(graph)\n",
    "rec_pipeline.graph.schema().view()\n",
    "\n",
    "print(\"--\")\n",
    "\n",
    "\n",
    "#  Nothing is output  (no schema is output; product bug ?)\n",
    "#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86588c83-e95e-4319-9956-3c22d23d2c18",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "##################################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "034dc9c8-d5c7-47ec-b2f4-1914931f5f60",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  this is mine\n",
    "#\n",
    "\n",
    "from tabulate import tabulate\n",
    "\n",
    "print(\"--\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c43a2f8-cd90-4c2e-af65-f41aa3a114b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  A bunch of diagnostics post load\n",
    "#\n",
    "\n",
    "#  List all node types with node counts\n",
    "\n",
    "l_query  = \"\"\"\n",
    "   MATCH (n) \n",
    "   WITH DISTINCT LABELS(n) AS temp, COUNT(n) AS tempCnt\n",
    "   UNWIND temp AS label\n",
    "   RETURN label, SUM(tempCnt) AS cnt\n",
    "   ORDER BY label\n",
    "   \"\"\".format()\n",
    "\n",
    "l_result = graph.query_unpaginated(l_query)\n",
    "   #\n",
    "print(tabulate(l_result, headers='keys', tablefmt='psql'))\n",
    "\n",
    "#  Sample output,\n",
    "#\n",
    "#     +----+-----------+-------+\n",
    "#     |    | label     |   cnt |\n",
    "#     |----+-----------+-------|\n",
    "#     |  0 | CELL_LINE |   494 |\n",
    "#     |  1 | DRUG      |   223 |\n",
    "#     |  2 | GDSC      |   266 |\n",
    "#     |  3 | GENE      |   719 |\n",
    "#     +----+-----------+-------+\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a5d9409-a4e0-49e6-8b41-67ba2dde5a3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  Return properties/keys from all vertices\n",
    "\n",
    "l_query  = \"\"\"\n",
    "   MATCH (n)\n",
    "   RETURN DISTINCT LABELS(n) AS label, KEYS(n)\n",
    "   \"\"\".format()\n",
    "\n",
    "l_result = graph.query_unpaginated(l_query)\n",
    "   #\n",
    "print(tabulate(l_result, headers='keys', tablefmt='psql'))\n",
    "\n",
    "#  Sample output,\n",
    "#\n",
    "#     +----+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+\n",
    "#     |    | label         | KEYS(n)                                                                                                                                                                                                                                                                                                                                                                                                                                          |\n",
    "#     |----+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|\n",
    "#     |  1 | ['GENE']      | ['id']                                                                                                                                                                                                                                                                                                                                                                                                                                           |\n",
    "#     |  2 | ['GDSC']      | ['id']                                                                                                                                                                                                                                                                                                                                                                                                                                           |\n",
    "#     |  7 | ['DRUG']      | ['Count', 'Name', 'PubCHEM', 'Sample Size', 'Synonyms', 'Target pathway', 'Targets', 'drug_embeddings', 'drug_id', 'id', 'smiles']                                                                                                                                                                                                                                                                                                               |\n",
    "#     | 11 | ['DRUG']      | ['Count', 'Name', 'PubCHEM', 'Sample Size', 'Target pathway', 'Targets', 'drug_embeddings', 'drug_id', 'id', 'smiles']                                                                                                                                                                                                                                                                                                                           |\n",
    "#     |  0 | ['CELL_LINE'] | ['CCLE_ID', 'Doubling.Time.Calculated.hrs', 'Gender', 'Hist_Subtype1', 'Hist_Subtype2', 'Hist_Subtype3', 'Histology', 'Name', 'PATHOLOGIST_ANNOTATION', 'Pathology', 'Site_Primary', 'Site_Subtype1', 'Site_Subtype2', 'Site_Subtype3', 'epigenomics_embeddings', 'genomics_embeddings', 'genomics_expression', 'genomics_methylation', 'genomics_mutation', 'id', 'mutRate', 'tcga_code', 'transcriptomics_embeddings', 'type', 'type_refined'] |\n",
    "#     |  3 | ['CELL_LINE'] | ['CCLE_ID', 'Doubling.Time.Calculated.hrs', 'epigenomics_embeddings', 'genomics_embeddings', 'genomics_expression', 'genomics_methylation', 'genomics_mutation', 'id', 'mutRate', 'tcga_code', 'transcriptomics_embeddings', 'type', 'type_refined']                                                                                                                                                                                             |\n",
    "#     |  4 | ['CELL_LINE'] | ['CCLE_ID', 'Gender', 'Hist_Subtype1', 'Hist_Subtype2', 'Hist_Subtype3', 'Histology', 'Name', 'PATHOLOGIST_ANNOTATION', 'Pathology', 'Site_Primary', 'Site_Subtype1', 'Site_Subtype2', 'Site_Subtype3', 'epigenomics_embeddings', 'genomics_embeddings', 'genomics_expression', 'genomics_methylation', 'genomics_mutation', 'id', 'mutRate', 'tcga_code', 'transcriptomics_embeddings', 'type', 'type_refined']                                 |\n",
    "#     |  5 | ['CELL_LINE'] | ['CCLE_ID', 'Doubling.Time.Calculated.hrs', 'Gender', 'Hist_Subtype1', 'Hist_Subtype2', 'Hist_Subtype3', 'Histology', 'Name', 'PATHOLOGIST_ANNOTATION', 'Pathology', 'Site_Primary', 'Site_Subtype1', 'Site_Subtype2', 'Site_Subtype3', 'epigenomics_embeddings', 'genomics_embeddings', 'genomics_expression', 'genomics_methylation', 'genomics_mutation', 'id', 'tcga_code', 'transcriptomics_embeddings', 'type', 'type_refined']            |\n",
    "#     |  6 | ['CELL_LINE'] | ['CCLE_ID', 'Doubling.Time.Calculated.hrs', 'Hist_Subtype1', 'Hist_Subtype2', 'Hist_Subtype3', 'Histology', 'Name', 'PATHOLOGIST_ANNOTATION', 'Pathology', 'Site_Primary', 'Site_Subtype1', 'Site_Subtype2', 'Site_Subtype3', 'epigenomics_embeddings', 'genomics_embeddings', 'genomics_expression', 'genomics_methylation', 'genomics_mutation', 'id', 'mutRate', 'tcga_code', 'transcriptomics_embeddings', 'type', 'type_refined']           |\n",
    "#     |  8 | ['CELL_LINE'] | ['CCLE_ID', 'Hist_Subtype1', 'Hist_Subtype2', 'Hist_Subtype3', 'Histology', 'Name', 'PATHOLOGIST_ANNOTATION', 'Pathology', 'Site_Primary', 'Site_Subtype1', 'Site_Subtype2', 'Site_Subtype3', 'epigenomics_embeddings', 'genomics_embeddings', 'genomics_expression', 'genomics_methylation', 'genomics_mutation', 'id', 'mutRate', 'tcga_code', 'transcriptomics_embeddings', 'type', 'type_refined']                                           |\n",
    "#     |  9 | ['CELL_LINE'] | ['CCLE_ID', 'Gender', 'Hist_Subtype1', 'Hist_Subtype2', 'Hist_Subtype3', 'Histology', 'Name', 'Pathology', 'Site_Primary', 'Site_Subtype1', 'Site_Subtype2', 'Site_Subtype3', 'epigenomics_embeddings', 'genomics_embeddings', 'genomics_expression', 'genomics_methylation', 'genomics_mutation', 'id', 'mutRate', 'tcga_code', 'transcriptomics_embeddings', 'type', 'type_refined']                                                           |\n",
    "#     | 10 | ['CELL_LINE'] | ['CCLE_ID', 'Doubling.Time.Calculated.hrs', 'Gender', 'Hist_Subtype1', 'Hist_Subtype2', 'Hist_Subtype3', 'Histology', 'Name', 'Pathology', 'Site_Primary', 'Site_Subtype1', 'Site_Subtype2', 'Site_Subtype3', 'epigenomics_embeddings', 'genomics_embeddings', 'genomics_expression', 'genomics_methylation', 'genomics_mutation', 'id', 'mutRate', 'tcga_code', 'transcriptomics_embeddings', 'type', 'type_refined']                           |\n",
    "#     | 12 | ['CELL_LINE'] | ['CCLE_ID', 'Hist_Subtype1', 'Hist_Subtype2', 'Hist_Subtype3', 'Histology', 'Name', 'Pathology', 'Site_Primary', 'Site_Subtype1', 'Site_Subtype2', 'Site_Subtype3', 'epigenomics_embeddings', 'genomics_embeddings', 'genomics_expression', 'genomics_methylation', 'genomics_mutation', 'id', 'mutRate', 'tcga_code', 'transcriptomics_embeddings', 'type', 'type_refined']                                                                     |\n",
    "#     +----+---------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13131bca-1ea1-4ddf-9e72-8090dc5c7519",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  List all vertice types with counts\n",
    "\n",
    "l_query  = \"\"\"\n",
    "   MATCH (m)-[r]->(n) \n",
    "   WITH DISTINCT TYPE(r) AS temp, COUNT(r) AS tempCnt\n",
    "   RETURN temp, tempCnt\n",
    "   ORDER BY temp\n",
    "   \"\"\".format()\n",
    "\n",
    "l_result = graph.query_unpaginated(l_query)\n",
    "   #\n",
    "print(tabulate(l_result, headers='keys', tablefmt='psql'))\n",
    "\n",
    "#  Sample output,\n",
    "#\n",
    "#     +----+----------------------+-----------+\n",
    "#     |    | temp                 |   tempCnt |\n",
    "#     |----+----------------------+-----------|\n",
    "#     |  0 | FOR_DRUG             |       238 |\n",
    "#     |  1 | HAS_CELL_LINE        |    105403 |\n",
    "#     |  2 | HAS_GENE_OBSERVATION |  18222436 |\n",
    "#     |  3 | SPLIT                |     94314 |\n",
    "#     +----+----------------------+-----------+\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb01a101-9312-40fb-aea1-9174d8069da4",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  List edge properties\n",
    "#\n",
    "\n",
    "l_query  = \"\"\"\n",
    "   MATCH (n) - [r] -> (m)\n",
    "   RETURN DISTINCT TYPE(r) AS my_type, KEYS(r) AS my_keys\n",
    "   \"\"\".format()\n",
    "\n",
    "l_result = graph.query_unpaginated(l_query)\n",
    "   #\n",
    "print(tabulate(l_result, headers='keys', tablefmt='psql'))\n",
    "\n",
    "#  Sample output,\n",
    "#\n",
    "#  +----+----------------------+-----------------------------------------------+\n",
    "#  |    | my_type              | my_keys                                       |\n",
    "#  |----+----------------------+-----------------------------------------------|\n",
    "#  |  0 | SPLIT                | ['label', 'split']                            |\n",
    "#  |  1 | HAS_CELL_LINE        | ['observation']                               |\n",
    "#  |  2 | FOR_DRUG             | []                                            |\n",
    "#  |  3 | HAS_GENE_OBSERVATION | ['full_column_name', 'observation', 'source'] |\n",
    "#  +----+----------------------+-----------------------------------------------+\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00042698-d382-4ed0-923f-a90fd1bd9954",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "##################################################\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ad1cca9-74b9-4e43-aa2e-aac854a9275b",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "start_time = default_timer()\n",
    "\n",
    "#  This comes from katana_pipeline.py\n",
    "#\n",
    "#  rec_pipeline.feature_generator()\n",
    "\n",
    "#  The above does this for 3 edge types; \n",
    "#     HAS_EXPRESSION_OBSERVATION\": \"genomics_expression\n",
    "#     HAS_METHYLATION_OBSERVATION\": \"genomics_methylation\n",
    "#     HAS_MUTATION_OBSERVATION\": \"genomics_mutation\n",
    "# \n",
    "#  By calling this,\n",
    "#     preprocessing.set_cell_line_property(self.graph, u, v)\n",
    "#\n",
    "#  Then it does this,,\n",
    "#       preprocessing.remove_null_cells(self.graph)\n",
    "#       preprocessing.delete_single_node(self.graph)\n",
    "\n",
    "\n",
    "##########################################################\n",
    "\n",
    "#  Listing source ..\n",
    "#\n",
    "\n",
    "#  We wont execute this here, and instead run below with checks pre and post \n",
    "#\n",
    "\n",
    "#  def set_cell_line_property(graph, edge_label: str, genomics_property: str):\n",
    "#     graph.query(\n",
    "#        f\"\"\"\n",
    "#        MATCH (c:CELL_LINE)-[r:{edge_label}]->(g:GENE)\n",
    "#        WITH c, r, g\n",
    "#        ORDER by g.ID\n",
    "#        WITH c, collect(r.observation) as {genomics_property}\n",
    "#        SET c.{genomics_property} = {genomics_property}\n",
    "#        \"\"\"\n",
    "#     )\n",
    "#  \n",
    "#  def delete_single_node(graph):\n",
    "#     \"\"\"Delete single node in a graph\"\"\"\n",
    "#     graph.query(\n",
    "#        \"\"\"\n",
    "#        MATCH (n)\n",
    "#        WHERE NOT (n)--()\n",
    "#        DELETE n\n",
    "#        \"\"\"\n",
    "#     )\n",
    "#  \n",
    "#  def remove_null_cells(graph):\n",
    "#     \"\"\"Delete cells node with NULL genomics features\"\"\"\n",
    "#     graph.query(\n",
    "#        \"\"\"\n",
    "#        MATCH (c:CELL_LINE)\n",
    "#        WHERE c.genomics_methylation IS NULL\n",
    "#           OR c.genomics_expression IS NULL\n",
    "#           OR c.genomics_mutation IS NULL\n",
    "#        DETACH DELETE c\n",
    "#        \"\"\"\n",
    "#     )\n",
    "#  \n",
    "#  print(f\"***Took {default_timer() - start_time} seconds to generate the features.***\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20b8936e-e862-4224-bfc7-12b4523f54ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  The above does this for 3 edge types; \n",
    "#     HAS_EXPRESSION_OBSERVATION\": \"genomics_expression\n",
    "#     HAS_METHYLATION_OBSERVATION\": \"genomics_methylation\n",
    "#     HAS_MUTATION_OBSERVATION\": \"genomics_mutation\n",
    "\n",
    "#  The above edge types do not exist\n",
    "#\n",
    "\n",
    "def set_cell_line_property(graph, edge_label: str, genomics_property: str):\n",
    "   l_return = graph.query_unpaginated(\n",
    "      f\"\"\"\n",
    "      \n",
    "      MATCH (c:CELL_LINE)-[r:{edge_label}]->(g:GENE)\n",
    "      WITH c, r, g\n",
    "      ORDER by g.ID\n",
    "      RETURN COUNT(*)\n",
    "      \n",
    "      // WITH c, collect(r.observation) as {genomics_property}\n",
    "      // SET c.{genomics_property} = {genomics_property}\n",
    "      \n",
    "      \"\"\"\n",
    "   )\n",
    "   print(l_return)\n",
    "\n",
    "set_cell_line_property(graph, \"HAS_EXPRESSION_OBSERVATION\", \"genomics_expression\")\n",
    "set_cell_line_property(graph, \"HAS_METHYLATION_OBSERVATION\", \"genomics_methylation\")\n",
    "set_cell_line_property(graph, \"HAS_MUTATION_OBSERVATION\", \"genomics_mutation\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c06a4184-b098-4c4a-a682-c65d154abc0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  Checking delete_single_node ..\n",
    "#\n",
    "\n",
    "#  One disconnected node is found, deleted\n",
    "#\n",
    "\n",
    "l_query  = \"\"\"\n",
    "\n",
    "   MATCH (n)\n",
    "   WHERE NOT (n)--()\n",
    "   \n",
    "   // RETURN COUNT(*)\n",
    "   //  Returns 1\n",
    "   \n",
    "   RETURN n\n",
    "   //  Returns\n",
    "   //\n",
    "   //     +----+-----------------+-------------+------------+--------------+\n",
    "   //     |    |           id(n) | labels(n)   | n.id       | n datatype   |\n",
    "   //     |----+-----------------+-------------+------------+--------------|\n",
    "   //     |  0 | 281474976710990 | ['GENE']    | Unnamed: 0 | node         |\n",
    "   //     +----+-----------------+-------------+------------+--------------+\n",
    "   \n",
    "   \"\"\".format()\n",
    "\n",
    "l_result = graph.query_unpaginated(l_query)\n",
    "   #\n",
    "print(tabulate(l_result, headers='keys', tablefmt='psql'))\n",
    "\n",
    "\n",
    "##########################################################\n",
    "\n",
    "#  Pre,\n",
    "#\n",
    "\n",
    "l_query  = \"\"\"\n",
    "   MATCH (n) \n",
    "   WITH DISTINCT LABELS(n) AS temp, COUNT(n) AS tempCnt\n",
    "   UNWIND temp AS label\n",
    "   RETURN label, SUM(tempCnt) AS cnt\n",
    "   ORDER BY label\n",
    "   \"\"\".format()\n",
    "\n",
    "l_result = graph.query_unpaginated(l_query)\n",
    "   #\n",
    "print(tabulate(l_result, headers='keys', tablefmt='psql'))\n",
    "\n",
    "#  Sample output,\n",
    "#\n",
    "#     +----+-----------+-------+\n",
    "#     |    | label     |   cnt |\n",
    "#     |----+-----------+-------|\n",
    "#     |  0 | CELL_LINE |   494 |\n",
    "#     |  1 | DRUG      |   223 |\n",
    "#     |  2 | GDSC      |   266 |\n",
    "#     |  3 | GENE      |   719 |\n",
    "#     +----+-----------+-------+\n",
    "\n",
    "##########################################################\n",
    "\n",
    "#  The actual method as written,\n",
    "#\n",
    "\n",
    "l_query  = \"\"\"\n",
    "\n",
    "   MATCH (n)\n",
    "   WHERE NOT (n)--()\n",
    "   DELETE n\n",
    "   \n",
    "   \"\"\".format()\n",
    "\n",
    "l_result = graph.query_unpaginated(l_query)\n",
    "   #\n",
    "print(tabulate(l_result, headers='keys', tablefmt='psql'))\n",
    "\n",
    "\n",
    "##########################################################\n",
    "\n",
    "#  Post,\n",
    "\n",
    "l_query  = \"\"\"\n",
    "   MATCH (n) \n",
    "   WITH DISTINCT LABELS(n) AS temp, COUNT(n) AS tempCnt\n",
    "   UNWIND temp AS label\n",
    "   RETURN label, SUM(tempCnt) AS cnt\n",
    "   ORDER BY label\n",
    "   \"\"\".format()\n",
    "\n",
    "l_result = graph.query_unpaginated(l_query)\n",
    "   #\n",
    "print(tabulate(l_result, headers='keys', tablefmt='psql'))\n",
    "\n",
    "#  Sample output,\n",
    "#\n",
    "#     +----+-----------+-------+\n",
    "#     |    | label     |   cnt |\n",
    "#     |----+-----------+-------|\n",
    "#     |  0 | CELL_LINE |   494 |\n",
    "#     |  1 | DRUG      |   223 |\n",
    "#     |  2 | GDSC      |   266 |\n",
    "#     |  3 | GENE      |   718 |\n",
    "#     +----+-----------+-------+\n",
    "\n",
    "print(\"--\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6029722a-afce-4407-ba4d-0e9f4baf85cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  Checking remove_null_cells\n",
    "#\n",
    "\n",
    "l_query  = \"\"\"\n",
    "\n",
    "   MATCH (c:CELL_LINE)\n",
    "   WHERE c.genomics_methylation IS NULL\n",
    "      OR c.genomics_expression IS NULL\n",
    "      OR c.genomics_mutation IS NULL\n",
    "   //DETACH DELETE c\n",
    "   RETURN COUNT(*)\n",
    "   \n",
    "   \"\"\".format()\n",
    "\n",
    "l_result = graph.query_unpaginated(l_query)\n",
    "   #\n",
    "print(tabulate(l_result, headers='keys', tablefmt='psql'))\n",
    "\n",
    "#  Sample results,\n",
    "#\n",
    "#     +----+------------+\n",
    "#     |    |   COUNT(*) |\n",
    "#     |----+------------|\n",
    "#     |  0 |          0 |\n",
    "#     +----+------------+\n",
    "\n",
    "#  Since this deletes nothing, we'll leave it be \n",
    "#\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30cc94fa-4b49-46c8-a6cc-3a4b9033b024",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  This is all read-only\n",
    "#\n",
    "\n",
    "stats = rec_pipeline.stats()\n",
    "stats\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de0a46ab-49e6-4961-8fcd-ffca4e0774f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc30ac93-1a4c-4099-b7e2-44949827d890",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "#  Same as above, we'll split this out line by line\n",
    "#\n",
    "#  The only effective code is split_generator, which is listed below\n",
    "#\n",
    "\n",
    "#  start_time = default_timer()\n",
    "#  \n",
    "#  rec_pipeline.split_generator(input_config)\n",
    "#  \n",
    "#  print(f\"***Took {default_timer() - start_time} seconds to generate the split.***\")\n",
    "#  \n",
    "#  rec_pipeline.graph.schema().view()\n",
    "\n",
    "\n",
    "##########################################################\n",
    "\n",
    "\n",
    "#  From katana_pipeline.py\n",
    "#\n",
    "\n",
    "    @utils.disable_warnings\n",
    "    def split_generator(self, input_hp) -> None:\n",
    "        \"\"\"Add split edges in the graph\n",
    "\n",
    "        Args:\n",
    "            input_hp: The graph input parameters.\n",
    "        \"\"\"\n",
    "\n",
    "        #  As expected, initially this does nothing, because there is no edge with the value \"SPLIT\"\n",
    "        #\n",
    "        print(\"Deleting old split\")\n",
    "        split.delete_split(self.graph)\n",
    " \n",
    "        #  We run this traversal below with a count\n",
    "        #\n",
    "        print(\"Generating split\")\n",
    "        split_df = pandas.concat(\n",
    "            self.graph.query(\n",
    "                \"\"\"\n",
    "            MATCH (a:DRUG)<-[:FOR_DRUG]-(:GDSC)-[u:HAS_CELL_LINE]->(c:CELL_LINE)\n",
    "            RETURN u.label as label, c.tcga_code as tcga_code,\n",
    "            a.id as drug_id, c.id as cell_line_id\n",
    "            \"\"\"\n",
    "            )\n",
    "        )\n",
    "        split_df = split_df.sort_values(by=[\"drug_id\", \"cell_line_id\", \"label\"]).reset_index(drop=True)\n",
    "        df_split = split.generate_split(split_df, input_hp.random_state, input_hp.test_size)\n",
    "        print(\"Upsert split on the graph\")\n",
    "        split.upsert_split(self.graph, df_split)\n",
    "        \n",
    "        \n",
    "##########################################################\n",
    "\n",
    "\n",
    "#  From split.py\n",
    "#\n",
    "\n",
    "def delete_split(graph):\n",
    "    graph.query(\n",
    "        \"\"\"\n",
    "        MATCH (:DRUG)-[u:SPLIT]->(:CELL_LINE)\n",
    "        DELETE u\n",
    "        \"\"\"\n",
    "    )\n",
    "\n",
    "def generate_split(df, random_state, test_size):\n",
    "    df_train, df_test = train_test_split( df, test_size=test_size, random_state=random_state, stratify=df[\"tcga_code\"], shuffle=True, )\n",
    "    df_train[\"split\"], df_test[\"split\"] = SplitType.TRAIN, SplitType.VAL\n",
    "    df_split = pandas.concat([df_train, df_test], ignore_index=True)\n",
    "    df_split = df_split.astype({\"drug_id\": \"string\", \"cell_line_id\": \"string\"})\n",
    "    return df_split.drop(columns=[\"tcga_code\"])\n",
    "\n",
    "def upsert_split(graph, split):\n",
    "    with DataFrameImporter(graph) as df_importer:\n",
    "        df_importer.node_id_property_name(\"id\")\n",
    "        df_importer.edges_dataframe(\n",
    "            split,\n",
    "            source_id_space=\"DRUG\",\n",
    "            destination_id_space=\"CELL_LINE\",\n",
    "            source_column=\"drug_id\",\n",
    "            destination_column=\"cell_line_id\",\n",
    "            type=\"SPLIT\",\n",
    "        )\n",
    "        df_importer.insert()\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e54e17f-4839-4a41-82f6-aae46e3bfa72",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  Testing this code, running a count instead\n",
    "#\n",
    "\n",
    "#       split_df = pandas.concat(\n",
    "#           self.graph.query(\n",
    "#               \"\"\"\n",
    "#           MATCH (a:DRUG)<-[:FOR_DRUG]-(:GDSC)-[u:HAS_CELL_LINE]->(c:CELL_LINE)\n",
    "#           RETURN u.label as label, c.tcga_code as tcga_code,\n",
    "#           a.id as drug_id, c.id as cell_line_id\n",
    "#           \"\"\"\n",
    "#           )\n",
    "#       )\n",
    "\n",
    "\n",
    "l_query  = \"\"\"\n",
    "\n",
    "   MATCH (a:DRUG)<-[:FOR_DRUG]-(:GDSC)-[u:HAS_CELL_LINE]->(c:CELL_LINE)\n",
    "   RETURN COUNT(*)\n",
    "   // RETURN u.label as label, c.tcga_code as tcga_code,\n",
    "   //    a.id as drug_id, c.id as cell_line_id\n",
    "   \n",
    "   \"\"\".format()\n",
    "\n",
    "l_result = graph.query_unpaginated(l_query)\n",
    "   #\n",
    "print(tabulate(l_result, headers='keys', tablefmt='psql'))\n",
    "\n",
    "#  Sample output\n",
    "#\n",
    "#     +----+------------+\n",
    "#     |    |   COUNT(*) |\n",
    "#     |----+------------|\n",
    "#     |  0 |      94314 |\n",
    "#     +----+------------+\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "64ef1bfd-b474-4ace-8991-e29ea2c27a86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "86d1ef8477c1462da611d6e94a2dcf3e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "          0/? [?op/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "94314\n",
      "+----+---------+-------------+-----------+----------------+\n",
      "|    | label   | tcga_code   |   drug_id | cell_line_id   |\n",
      "|----+---------+-------------+-----------+----------------|\n",
      "|  0 |         | OV          |  10027278 | ACH-000001     |\n",
      "|  1 |         | LAML        |  10027278 | ACH-000002     |\n",
      "|  2 |         | LAML        |  10027278 | ACH-000004     |\n",
      "|  3 |         | LAML        |  10027278 | ACH-000006     |\n",
      "|  4 |         | COAD/READ   |  10027278 | ACH-000007     |\n",
      "+----+---------+-------------+-----------+----------------+\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import pandas\n",
    "\n",
    "#  So we'll actually run the above\n",
    "#\n",
    "\n",
    "# split_df = pandas.concat(\n",
    "split_df =  graph.query_unpaginated(\n",
    "   \"\"\"\n",
    "   \n",
    "   MATCH (a:DRUG)<-[:FOR_DRUG]-(:GDSC)-[u:HAS_CELL_LINE]->(c:CELL_LINE)\n",
    "   \n",
    "   RETURN u.label as label, c.tcga_code as tcga_code, a.id as drug_id, c.id as cell_line_id\n",
    "   \"\"\"\n",
    "   )\n",
    "# )\n",
    "\n",
    "print(len(split_df))\n",
    "    \n",
    "split_df = split_df.sort_values(by=[\"drug_id\", \"cell_line_id\", \"label\"]).reset_index(drop=True)\n",
    "\n",
    "print(tabulate(split_df.head(5), headers='keys', tablefmt='psql'))\n",
    "\n",
    "\n",
    "#  df_split = split.generate_split(split_df, input_hp.random_state, input_hp.test_size)\n",
    "\n",
    "\n",
    "#  print(\"Upsert split on the graph\")\n",
    "#  split.upsert_split(self.graph, df_split)\n",
    "        \n",
    "        \n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58d96df1-4b23-4ff7-baa2-26d73a2fb748",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "540598be-a55e-4971-9212-c51d81dc50ff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e9feb12-a776-4a5f-90f1-5207598a86ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "933be6e4-cccf-4022-8dce-0231019d1f43",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b15d0d0-7686-4764-adc2-da340c754e03",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5535f50d-f4d3-4de0-a769-0cb62ac94cb5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a3b9391-ff8f-4d29-80a7-74ca610940ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d086603e-d2d4-45af-8afd-f91849e32f9a",
   "metadata": {
    "tags": []
   },
   "source": [
    "<a id='step-3'></a>\n",
    "## AI Training Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2df6ce75-84ea-4c86-9420-ea0d4f7668e9",
   "metadata": {},
   "source": [
    "<img src=\"images/pipeline_gnn.png\" style=\"width: 1500px;\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5c6bb1a-7c59-43eb-b5a2-d9c06183bf29",
   "metadata": {},
   "source": [
    "### Initialize model training"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e3de5b0-5f66-4dc6-8f30-3e66d789ecb7",
   "metadata": {},
   "source": [
    "<center><img src=\"./images/model.png\" style=\"width:70%;\"></center>\n",
    "\n",
    "DeepCDR is a hybrid graph convolutional network model that predicts the IC50 value, which denotes the effectiveness of a drug in inhibiting the growth of a specific cancer cell line.\n",
    "\n",
    "On the one hand, the graph neural network (GNN) takes the adjacent information of atoms in a drug into consideration by aggregating the features of neighboring atoms together. On the other hand, the subnetworks extract high-level features of cancer omics profiles from a certain cancer cell line (i.e., genomic data, transcriptomic data and epigenomic data). Then the high-level features of drug and multiple omics data get concatenated and fed to a 1-D\n",
    "convolutional neural network."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8394f5c-c3d8-4782-b34b-5d330c8c414a",
   "metadata": {},
   "source": [
    "### Load Model configuration\n",
    "\n",
    "The model configuration defines the specifications of the model that will be used to train on the graph.\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "You can change the model configuration in the <a href=\"./config/hyperparams.py\" class=\"alert-link\">hyperparams.py</a> file. By default the model is using genomics, epigenomics and transcriptomics data.\n",
    "</div>\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "You can change the model architecture in the <a href=\"./src/model.py\" class=\"alert-link\">model.py</a> file.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07e70ffb-3c7c-49ef-baa9-f4aff238e84e",
   "metadata": {},
   "outputs": [],
   "source": [
    "model_config = hyperparams.load_model_config()\n",
    "model_config"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "247df214-0c2a-4787-b97f-ca6d9b5f457f",
   "metadata": {},
   "source": [
    "### Load Training configuration\n",
    "\n",
    "The training configuration specify details of the training pipeline.\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "You can change the training configuration in the <a href=\"./config/hyperparams.py\" class=\"alert-link\">hyperparams.py</a> file.\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "171b3bae-5feb-482a-9baa-ecdbb8b1915e",
   "metadata": {},
   "outputs": [],
   "source": [
    "training_config = hyperparams.load_training_config()\n",
    "training_config"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c09dd293-1f51-4040-ba2b-44e130f67951",
   "metadata": {},
   "source": [
    "### Training\n",
    "\n",
    "Distributed training of the model. The trained model and its configuration will be saved in the `rec_pipeline` graph using the `experiment_name` graph property."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9faa3bf-3b43-4d26-a069-09f7bb342808",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = default_timer()\n",
    "validation_metric = rec_pipeline.train(model_config, training_config)\n",
    "print(f\"***Took {default_timer() - start_time} seconds to train the model.***\")\n",
    "print(\"Validation metric: \", validation_metric)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1350f3c1-27aa-40e6-811d-7a57a05abdea",
   "metadata": {},
   "source": [
    "### Testing\n",
    "\n",
    "Test the trained model on the test data and compare the results with other achitecture baselines. We compare three different metrics:\n",
    "- [Pearson correlation coefficient](https://en.wikipedia.org/wiki/Pearson_correlation_coefficient): a measure of linear correlation between two sets of data. A value closer to 1 is better.\n",
    "- [Spearman correlation coefficient](https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient): how well the relationship between two variables can be described using a monotonic function. A value closer to 1 is better.\n",
    "- [Root mean square error (RMSE)](https://en.wikipedia.org/wiki/Root-mean-square_deviation): a measure of the differences between values predicted by a model and the values observed. A lower value is better."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b64cb014-1302-4bac-b55f-3c86ba61436c",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = default_timer()\n",
    "metrics, predictions, labels = rec_pipeline.test(training_config)\n",
    "print(f\"***Took {default_timer() - start_time} seconds to test the model.***\")\n",
    "metrics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ce573cf-65ab-4c2c-ba13-8cf30d4e595a",
   "metadata": {},
   "source": [
    "### Plot Predictions\n",
    "\n",
    "Create a scatter plot of the `IC50` predicted by the trained model in function of `IC50` real value."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ec43d5c-07d9-456c-9aa6-c3d230b64991",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "start_time = default_timer()\n",
    "rec_pipeline.plot(labels, predictions)\n",
    "print(f\"***Took {default_timer() - start_time} seconds to plot figures.***\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1411c92-b3ca-4340-94d6-2f3296607c31",
   "metadata": {},
   "source": [
    "<a id='step-4'></a>\n",
    "## Trained model inference "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50716925-45aa-426b-adde-f79946d387ec",
   "metadata": {},
   "source": [
    "<img src=\"images/pipeline_inference.png\" style=\"width: 1500px;\"/>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edb4d9c2-ccda-4ce1-b480-a109e492ba58",
   "metadata": {},
   "outputs": [],
   "source": [
    "rec_pipeline.infer(training_config)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e46e3750-0856-4a72-a332-a8bab0a7230a",
   "metadata": {},
   "outputs": [],
   "source": [
    "bortezomib = \"B(C(CC(C)C)NC(=O)C(CC1=CC=CC=C1)NC(=O)C2=NC=CN=C2)(O)O\"\n",
    "cell_line = \"ACH-000001\"\n",
    "rec_pipeline.infer(training_config, drug=bortezomib, cell_line=cell_line)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97320404-c952-4b8d-94e9-ba6424c155ce",
   "metadata": {},
   "source": [
    "## Run trained model to save node embeddings\n",
    "\n",
    "Save the trained model embeddings as a node property. `drug_embeddings`, `epigenomics_embeddings`, `genomics_embeddings` and `transcriptomics_embeddings` will be created and saved. Those embeddings can be use for a future downstream task with a different model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fadc229-457e-40f5-a0df-b1fd36a772ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = default_timer()\n",
    "rec_pipeline.infer_embeddings(model_config)\n",
    "print(f\"***Took {default_timer() - start_time} seconds to save node embeddings.***\")\n",
    "rec_pipeline.graph.schema().view()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e058913-fc3e-4c06-888b-b7fd9e167698",
   "metadata": {},
   "source": [
    "## Save Graph\n",
    "\n",
    "Save the graph created to the bucket location named in the `save_graph_path` variable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36cc1319-b49e-4f07-aef0-a140ff354344",
   "metadata": {},
   "outputs": [],
   "source": [
    "if input_config.save_graph_path:\n",
    "    export_data.rdg(graph, input_config.save_graph_path)"
   ]
  }
 ],
 "metadata": {
  "katana": {
   "component": "AI",
   "max_num_hosts": 4
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "toc-autonumbering": false
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
