{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57a6c20f-3419-4768-92a0-f106c5931689",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  Notebook to create and load the DeepCDR graph ..\n",
    "#\n",
    "#  Be advised;\n",
    "#\n",
    "#     .  We load the graph using Pandas dataframes from the Jupyter container.\n",
    "#        This is about the slowest way to do this.\n",
    "#        We use this approach because of ease; we do not need to host the data\n",
    "#        files to load, create and use authorization tokens, etc.\n",
    "#\n",
    "#        In production (or to run faster) we should load using Dask dataframes.\n",
    "#\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "212d8ad3-a876-4638-b95e-a0702f897587",
   "metadata": {},
   "source": [
    "#  Setup: Display options .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a38ebd7-d1d9-47a8-955c-a8270b024543",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "%xmode Minimal\n",
    "\n",
    "#  Setting display options \n",
    "\n",
    "import pandas as pd\n",
    "import numpy  as np\n",
    "   #\n",
    "pd.set_option(\"display.width\", 480)\n",
    "\n",
    "#  Sets horizontal scroll for wide outputs\n",
    "#\n",
    "from IPython.display import display, HTML\n",
    "display(HTML(\"\"))\n",
    "\n",
    "from tabulate import tabulate\n",
    "\n",
    "\n",
    "print(\"--\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ac9bc73-49ab-44bf-a21a-a2a589622a51",
   "metadata": {},
   "source": [
    "#  Setup: Connectivity .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e87b75ba-c072-4111-8cca-4b4f394a0fce",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from katana import remote\n",
    "\n",
    "my_client = remote.Client()\n",
    "\n",
    "print(my_client)\n",
    "print(my_client.server_version)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3a77f99-f11b-4105-ab43-17ce605b15e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "NUM_PARTITIONS  = 3\n",
    "   #\n",
    "DB_NAME         = \"my_db\"\n",
    "GRAPH_NAME      = \"my_graph\"\n",
    "\n",
    "print(\"--\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73f507b5-9a82-4441-a79f-cc2fc239968a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ##################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f2292b8-7df0-49a0-b3da-523f56cfe4f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  DELETE ALL GRAPHS\n",
    "\n",
    "for l_database in my_client.databases():\n",
    "   for l_graph in my_client.get_database(name = l_database.name).graphs_in_database():\n",
    "      l_handle = my_client.get_database(name = l_database.name).get_graph_by_id(id = l_graph.graph_id)\n",
    "      l_handle.delete()\n",
    "\n",
    "for l_graph in my_client.graphs():\n",
    "   print(\"GRAPH ID: \", l_graph.graph_id, \"      GRAPH Version: \", l_graph.version)\n",
    "\n",
    "print(\"--\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "c989fda5-3ce8-4589-bf36-4da42d3d993c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GRAPH ID:  5NUwpV8GmVPB2FWCGYwWc32GQrPySTb38Cw7A29kQTvx    GRAPH Version:  1    Graph Name:  my_graph_BC\n",
      "GRAPH ID:  6GYEDQxA2ewmmGV53Vb7GDBx5U9osUzBctJhh6auAdG3    GRAPH Version:  97    Graph Name:  my_graph_BB\n",
      "GRAPH ID:  7moQEUJpTBLCoLgABUjjAyFaSMMdanvqSmde4HBL9qpX    GRAPH Version:  2    Graph Name:  my_graph_BD\n",
      "--\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#  DELETE ONE SPECIFIC GRAPH ONLY\n",
    "\n",
    "for l_database in my_client.databases():\n",
    "   for l_graph in my_client.get_database(name = l_database.name).graphs_in_database():\n",
    "      if ((l_database.name == DB_NAME) and (l_graph.name == GRAPH_NAME)):   \n",
    "         l_handle = my_client.get_database(name = l_database.name).get_graph_by_id(id = l_graph.graph_id)\n",
    "         l_handle.delete()\n",
    "\n",
    "for l_graph in my_client.graphs():\n",
    "   print(\"GRAPH ID: \", l_graph.graph_id, \"   GRAPH Version: \", l_graph.version, \"   Graph Name: \", l_graph.name)\n",
    "\n",
    "print(\"--\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71f69a0e-451b-4679-b49c-5db1208d4c8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  DELETE ALL DATABASES\n",
    "\n",
    "for l_database in my_client.databases():\n",
    "   if (l_database.name != \"default\"):\n",
    "      my_client.get_database(name = l_database.name).delete_database()\n",
    "      print(\"--\")\n",
    "\n",
    "for l_database in my_client.databases():\n",
    "   print(\"DB ID: \", l_database.database_id, \"     DB Name: \", l_database.name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3b3bc00-fafb-481d-ba5a-8b52ae7d6e79",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  CREATE DATABASE\n",
    "\n",
    "my_database = my_client.create_database(name = DB_NAME)\n",
    "\n",
    "print(my_database.database_id)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "b8ba088b-2ec5-4b2a-8171-3c87ddffb47c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<_Graph my_graph, EfDAPbeB2xFpJtayrQpM2gxN42vpPcPhWeHBiQvr2Yra, 0>\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#  CREATE GRAPH\n",
    "#\n",
    "my_graph = my_client.get_database(name = DB_NAME).create_graph(name = GRAPH_NAME, num_partitions = NUM_PARTITIONS)\n",
    "\n",
    "print(my_graph)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3217f3b-dd84-47a3-ae2a-5033d122604a",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  CONNECT TO GRAPH\n",
    "\n",
    "my_graph, *_ = my_client.get_database(name=DB_NAME).find_graphs_by_name(GRAPH_NAME)\n",
    "\n",
    "print(my_graph)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "918521c4-3808-4d15-a09d-8417964f5fb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ##################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e576c04e-9cd8-41e0-a1a3-ae09397e4f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  Now that we share servers, get a better list of databases and graphs ..\n",
    "#\n",
    "\n",
    "for l_each_d in my_client.databases():\n",
    "   print(\"Database Name: %-28s   Id: %s\" % (l_each_d.name, l_each_d.database_id))\n",
    "      #\n",
    "   for l_each_g in my_client.get_database(name=l_each_d.name).graphs_in_database(): \n",
    "      print(\"   Graph Name: %-28s      Version: %-8d   Id: %s\" % (l_each_g.name, l_each_g.version, l_each_g.graph_id))\n",
    "\n",
    "\n",
    "print(\"\"  )\n",
    "print(\"--\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d1dee0a-411c-4269-95b3-9cb5ded64676",
   "metadata": {},
   "source": [
    "#  Step 01: Process CSV files .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1af9c83f-ba62-4757-bd8d-d74c49d4e7c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  Read the CSVs as Step 01 towards populating our graph\n",
    "#\n",
    "\n",
    "\n",
    "l_folder   = \"./10_Data/\"\n",
    "\n",
    "\n",
    "#  Here's the schema and sample data for the node files,\n",
    "#\n",
    "#     ==> 21_drug.txt <==\n",
    "#     id|smiles|label\n",
    "#     10027278|NCC(=O)Nc1ccc(-n2nc(C(F)(F)F)cc2-c2ccc3c(ccc4ccccc43)c2)cc1|DRUG\n",
    "#     \n",
    "#     ==> 22_gdsc.txt <==\n",
    "#     id|label\n",
    "#     GDSC:1|GDSC\n",
    "#     \n",
    "#     ==> 23_cell_line.txt <==\n",
    "#     id|tcga_code|label\n",
    "#     ACH-000001|OV|CELL_LINE\n",
    "#     \n",
    "#     ==> 24_gene.txt <==\n",
    "#     id\n",
    "#     AKT2\n",
    "\n",
    "#  There is null data in CELL_LINE.label\n",
    "#\n",
    "\n",
    "l_nodes    = [\n",
    "   { \"file\": \"21_drug.txt.gz\"      , \"label\": \"DRUG\"      },\n",
    "   { \"file\": \"22_gdsc.txt.gz\"      , \"label\": \"GDSC\"      },\n",
    "   { \"file\": \"23_cell_line.txt.gz\" , \"label\": \"CELL_LINE\" },\n",
    "   { \"file\": \"24_gene.txt.gz\"      , \"label\": \"GENE\"      },\n",
    "   ]\n",
    "\n",
    "pd_nodes = {}\n",
    "   #\n",
    "print(\"Just Nodes ..\")\n",
    "print()\n",
    "   #\n",
    "for l_each in l_nodes:\n",
    "   pd_nodes[l_each[\"label\"]] = pd.read_csv( (l_folder + l_each[\"file\"]), header = \"infer\", sep=\"|\",\n",
    "       dtype={\"id\": \"string\", \"smiles\": \"string\", \"label\": \"string\", \"tcga_code\": \"string\"})\n",
    "   print(\"Number of records: %-8d   %-16s   %-32s\" % (len(pd_nodes[l_each[\"label\"]]), l_each[\"label\"], l_each[\"file\"]))\n",
    "   print(tabulate(pd_nodes[l_each[\"label\"]].head(2), headers='keys', tablefmt='psql', showindex=False))\n",
    "   print()\n",
    "\n",
    "    \n",
    "   ######################################################### \n",
    "    \n",
    "    \n",
    "#  Here's the schema and sample data for the first group of edge files,\n",
    "#\n",
    "#     ==> 31_gdsc_drug.txt <==\n",
    "#     START_ID|END_ID|TYPE\n",
    "#     GDSC:1|176870|FOR_DRUG\n",
    "#     \n",
    "#     ==> 32_gdsc_cell_line.txt <==\n",
    "#     START_ID|END_ID|label|TYPE\n",
    "#     GDSC:1|ACH-002137||HAS_CELL_LINE\n",
    "#     \n",
    "#     ==> 33_cell_line_gene_expression.txt <==\n",
    "#     START_ID|END_ID|observation|TYPE\n",
    "#     ACH-000828|LASP1|9.39347626575|HAS_EXPRESSION_OBSERVATION\n",
    "#     \n",
    "#     ==> 34_cell_line_gene_methylation.txt <==\n",
    "#     START_ID|END_ID|observation|TYPE\n",
    "#     ACH-000001|ABL1|0.00153|HAS_METHYLATION_OBSERVATION\n",
    "\n",
    "l_edges1   = [\n",
    "   { \"file\": \"31_gdsc_drug.txt.gz\"                 , \"type\": \"FOR_DRUG\"                   },\n",
    "   { \"file\": \"32_gdsc_cell_line.txt.gz\"            , \"type\": \"HAS_CELL_LINE\"              },                       #  Alternately called \"HAS\" and \"FOR\"\n",
    "   { \"file\": \"33_cell_line_gene_expression.txt.gz\" , \"type\": \"HAS_EXPRESSION_OBSERVATION\" },\n",
    "   { \"file\": \"34_cell_line_gene_methylation.txt.gz\", \"type\": \"HAS_METHYLATION_OBSERVATION\"},\n",
    "   ]\n",
    "\n",
    "pd_edges1 = {}\n",
    "   #\n",
    "print()\n",
    "print(\"Just Edges group 1 ..\")\n",
    "print()\n",
    "   #\n",
    "for l_each in l_edges1:\n",
    "   pd_edges1[l_each[\"type\"]] = pd.read_csv( (l_folder + l_each[\"file\"]), header = \"infer\", sep=\"|\",\n",
    "      dtype={\"START_ID\": \"string\", \"END_ID\": \"string\", \"TYPE\": \"string\", \"label\": \"string\", \"observation\": \"float64\"})\n",
    "   print(\"Number of records: %-8d   %-32s   %-32s\" % (len(pd_edges1[l_each[\"type\"]]), l_each[\"type\"], l_each[\"file\"]))\n",
    "   print(tabulate(pd_edges1[l_each[\"type\"]].head(2), headers='keys', tablefmt='psql', showindex=False))\n",
    "   print()\n",
    "    \n",
    "    \n",
    "   ######################################################### \n",
    "\n",
    "\n",
    "#  Sample data,\n",
    "#\n",
    "#     START_ID|END_ID|observation|mutation|TYPE\n",
    "#     ACH-000381|FH|0.0|FH.1:241671906|HAS_MUTATION_OBSERVATION\n",
    "\n",
    "l_edges2   = [ 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58 ,59, 60, 61, 62, 63, 64, 65 ]\n",
    "   #\n",
    "pd_edges2 = {}\n",
    "\n",
    "print()\n",
    "print(\"Just Edges group 2 ..\")\n",
    "print()\n",
    "   #\n",
    "for l_each in l_edges2:\n",
    "   l_file = (str(l_each) + \"_cell_line_gene_mutation.txt.gz\")\n",
    "      #\n",
    "   pd_edges2[str(l_each)] = pd.read_csv( (l_folder + l_file), header = \"infer\", sep=\"|\",\n",
    "      dtype={\"START_ID\": \"string\", \"END_ID\": \"string\", \"TYPE\": \"string\", \"mutation\": \"string\", \"observation\": \"float64\"})\n",
    "   print(\"Number of records: %-8d   %-32s   %-32s\" % (len(pd_edges2[str(l_each)]), \"HAS_MUTATION_OBSERVATION\", l_file))\n",
    "   #  print(tabulate(pd_edges2[str(l_each)].head(2), headers='keys', tablefmt='psql', showindex=False))\n",
    "   #  print()\n",
    "\n",
    "print()\n",
    "print(\"Concatenating ..\")\n",
    "print()\n",
    "   #\n",
    "pd_edges2_all = pd.concat(pd_edges2).droplevel(0).reindex()                                          #  The concat() will give us a multi-level index\n",
    "pd_edges2     = {}                                                                                   #  which Dask, and thus the graph importer, will\n",
    "   #\n",
    "print(\"Number of records: %-8d   %-32s\" % (len(pd_edges2_all), \"HAS_MUTATION_OBSERVATION\"))\n",
    "print(tabulate(pd_edges2_all.head(2), headers='keys', tablefmt='psql', showindex=False))\n",
    "print()\n",
    "\n",
    "print(\"--\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "00c5ead0-8e50-47f9-95d1-2a7b426d1712",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "#  Step 00: Sample output from above .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14641822-29e5-457b-8182-a7e384b0ffdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#  Sample output,\n",
    "#\n",
    "#     Just Nodes ..\n",
    "#     \n",
    "#     Number of records: 223        DRUG               21_drug.txt.gz                  \n",
    "#     +----------+-------------------------------------------------------------+---------+\n",
    "#     |       id | smiles                                                      | label   |\n",
    "#     |----------+-------------------------------------------------------------+---------|\n",
    "#     | 10027278 | NCC(=O)Nc1ccc(-n2nc(C(F)(F)F)cc2-c2ccc3c(ccc4ccccc43)c2)cc1 | DRUG    |\n",
    "#     | 10074640 | Cc1ccc(NC(=O)c2ccc(CN3CCN(C)CC3)cc2)cc1Nc1nc(-c2cccnc2)cs1  | DRUG    |\n",
    "#     +----------+-------------------------------------------------------------+---------+\n",
    "#     \n",
    "#     Number of records: 266        GDSC               22_gdsc.txt.gz                  \n",
    "#     +-----------+---------+\n",
    "#     | id        | label   |\n",
    "#     |-----------+---------|\n",
    "#     | GDSC:1    | GDSC    |\n",
    "#     | GDSC:1001 | GDSC    |\n",
    "#     +-----------+---------+\n",
    "#     \n",
    "#     Number of records: 1457       CELL_LINE          23_cell_line.txt.gz             \n",
    "#     +------------+-------------+-----------+\n",
    "#     | id         | tcga_code   | label     |\n",
    "#     |------------+-------------+-----------|\n",
    "#     | ACH-000001 | OV          | CELL_LINE |\n",
    "#     | ACH-000002 | LAML        | CELL_LINE |\n",
    "#     +------------+-------------+-----------+\n",
    "#     \n",
    "#     Number of records: 719        GENE               24_gene.txt.gz                  \n",
    "#     +-------+\n",
    "#     | id    |\n",
    "#     |-------|\n",
    "#     | AKT2  |\n",
    "#     | EPAS1 |\n",
    "#     +-------+\n",
    "#     \n",
    "#     \n",
    "#     Just Edges group 1 ..\n",
    "#     \n",
    "#     Number of records: 266        FOR_DRUG                           31_gdsc_drug.txt.gz             \n",
    "#     +------------+----------+----------+\n",
    "#     | START_ID   |   END_ID | TYPE     |\n",
    "#     |------------+----------+----------|\n",
    "#     | GDSC:1     |   176870 | FOR_DRUG |\n",
    "#     | GDSC:1001  |    65110 | FOR_DRUG |\n",
    "#     +------------+----------+----------+\n",
    "#     \n",
    "#     Number of records: 257754     FOR_CELL_LINE                      32_gdsc_cell_line.txt.gz        \n",
    "#     +------------+------------+---------+---------------+\n",
    "#     | START_ID   | END_ID     | label   | TYPE          |\n",
    "#     |------------+------------+---------+---------------|\n",
    "#     | GDSC:1     | ACH-002137 | <NA>    | HAS_CELL_LINE |\n",
    "#     | GDSC:1     | ACH-000474 | <NA>    | HAS_CELL_LINE |\n",
    "#     +------------+------------+---------+---------------+\n",
    "#     \n",
    "#     Number of records: 391017     HAS_EXPRESSION_OBSERVATION         33_cell_line_gene_expression.txt.gz\n",
    "#     +------------+----------+---------------+----------------------------+\n",
    "#     | START_ID   | END_ID   |   observation | TYPE                       |\n",
    "#     |------------+----------+---------------+----------------------------|\n",
    "#     | ACH-000828 | LASP1    |     9.39348   | HAS_EXPRESSION_OBSERVATION |\n",
    "#     | ACH-000828 | HOXA11   |     0.0426443 | HAS_EXPRESSION_OBSERVATION |\n",
    "#     +------------+----------+---------------+----------------------------+\n",
    "#     \n",
    "#     Number of records: 426500     HAS_METHYLATION_OBSERVATION        34_cell_line_gene_methylation.txt.gz\n",
    "#     +------------+----------+---------------+-----------------------------+\n",
    "#     | START_ID   | END_ID   |   observation | TYPE                        |\n",
    "#     |------------+----------+---------------+-----------------------------|\n",
    "#     | ACH-000001 | ABL1     |       0.00153 | HAS_METHYLATION_OBSERVATION |\n",
    "#     | ACH-000001 | ABL1     |       0.00591 | HAS_METHYLATION_OBSERVATION |\n",
    "#     +------------+----------+---------------+-----------------------------+\n",
    "#     \n",
    "#     \n",
    "#     Just Edges group 2 ..\n",
    "#     \n",
    "#     Number of records: 500000     HAS_MUTATION_OBSERVATION           41_cell_line_gene_mutation.txt.gz\n",
    "#     Number of records: 500000     HAS_MUTATION_OBSERVATION           42_cell_line_gene_mutation.txt.gz\n",
    "#        ...\n",
    "#     Number of records: 500000     HAS_MUTATION_OBSERVATION           63_cell_line_gene_mutation.txt.gz\n",
    "#     Number of records: 500000     HAS_MUTATION_OBSERVATION           64_cell_line_gene_mutation.txt.gz\n",
    "#     Number of records: 422852     HAS_MUTATION_OBSERVATION           65_cell_line_gene_mutation.txt.gz\n",
    "#     \n",
    "#     Concatenating ..\n",
    "#     \n",
    "#     Number of records: 12422852   HAS_MUTATION_OBSERVATION        \n",
    "#     +------------+----------+---------------+--------------------+--------------------------+\n",
    "#     | START_ID   | END_ID   |   observation | mutation           | TYPE                     |\n",
    "#     |------------+----------+---------------+--------------------+--------------------------|\n",
    "#     | ACH-000828 | TNFRSF14 |             0 | TNFRSF14.1:2488170 | HAS_MUTATION_OBSERVATION |\n",
    "#     | ACH-000828 | TNFRSF14 |             0 | TNFRSF14.1:2489805 | HAS_MUTATION_OBSERVATION |\n",
    "#     +------------+----------+---------------+--------------------+--------------------------+\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31d0d07d-6936-4dd8-8c08-6e5b9e29639a",
   "metadata": {},
   "source": [
    "#  Step 02:  Create/Load graph .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3271fdaa-0f21-4f75-b1b3-b8996ff3ea84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c40de14137284080a44f5ac3ccdccca1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "          0/? [?op/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3cabdc2a4f274d71a3b2efca97eccb04",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "          0/? [?op/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "This last dataframe is large, and takes a many minutes\n",
      "before you will see any output/progress ..\n",
      "\n",
      "Further, we are loading using Pandas dataframes from the\n",
      "Jupyter container, which is the slowest choice. For speed,\n",
      "we should be using Dask dataframes loaded from files\n",
      "accessible from the KGIP worker nodes.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "#  At this point we have several Pandas DataFrames; import them into the graph-\n",
    "#\n",
    "#     Why use Pandas DataFrames versus Dask ?\n",
    "#     \n",
    "#     .  The data is small enough\n",
    "#     .  Dask DataFrames currently (beta) have to be sourced from a shared/public drive\n",
    "#\n",
    "\n",
    "\n",
    "from katana.remote import import_data\n",
    "\n",
    "\n",
    "#  Just nodes\n",
    "#\n",
    "with import_data.DataFrameImporter(my_graph) as df_importer:   \n",
    "   df_importer.nodes_dataframe(\n",
    "      pd_nodes[\"DRUG\"],\n",
    "      id_column             = \"id\",\n",
    "      id_space              = \"DRUG\",  \n",
    "      label                 = \"DRUG\",  \n",
    "      ) \n",
    "   df_importer.nodes_dataframe(\n",
    "      pd_nodes[\"GDSC\"],\n",
    "      id_column             = \"id\",\n",
    "      id_space              = \"GDSC\",  \n",
    "      label                 = \"GDSC\",  \n",
    "      ) \n",
    "   df_importer.nodes_dataframe(\n",
    "      pd_nodes[\"CELL_LINE\"],\n",
    "      id_column             = \"id\",\n",
    "      id_space              = \"CELL_LINE\",  \n",
    "      label                 = \"CELL_LINE\",  \n",
    "      ) \n",
    "   df_importer.nodes_dataframe(\n",
    "      pd_nodes[\"GENE\"],\n",
    "      id_column             = \"id\",\n",
    "      id_space              = \"GENE\",  \n",
    "      label                 = \"GENE\",  \n",
    "      ) \n",
    "   df_importer.insert()\n",
    "    \n",
    "               \n",
    "#  Just edges\n",
    "#\n",
    "with import_data.DataFrameImporter(my_graph) as df_importer:   \n",
    "   df_importer.edges_dataframe(\n",
    "      pd_edges1[\"FOR_DRUG\"], \n",
    "      source_id_space       = \"GDSC\", \n",
    "      destination_id_space  = \"DRUG\",   \n",
    "      source_column         = \"START_ID\",\n",
    "      destination_column    = \"END_ID\",\n",
    "      type                  = \"FOR_DRUG\"\n",
    "      )\n",
    "   df_importer.edges_dataframe(\n",
    "      pd_edges1[\"HAS_CELL_LINE\"], \n",
    "      source_id_space       = \"GDSC\", \n",
    "      destination_id_space  = \"CELL_LINE\",   \n",
    "      source_column         = \"START_ID\",\n",
    "      destination_column    = \"END_ID\",\n",
    "      type                  = \"HAS_CELL_LINE\"                        #  Alternately called \"HAS\" and \"FOR\"\n",
    "      )\n",
    "   df_importer.edges_dataframe(\n",
    "      pd_edges1[\"HAS_EXPRESSION_OBSERVATION\"], \n",
    "      source_id_space       = \"CELL_LINE\", \n",
    "      destination_id_space  = \"GENE\",   \n",
    "      source_column         = \"START_ID\",\n",
    "      destination_column    = \"END_ID\",\n",
    "      type                  = \"HAS_EXPRESSION_OBSERVATION\"\n",
    "      )\n",
    "   df_importer.edges_dataframe(\n",
    "      pd_edges1[\"HAS_METHYLATION_OBSERVATION\"], \n",
    "      source_id_space       = \"CELL_LINE\", \n",
    "      destination_id_space  = \"GENE\",   \n",
    "      source_column         = \"START_ID\",\n",
    "      destination_column    = \"END_ID\",\n",
    "      type                  = \"HAS_METHYLATION_OBSERVATION\"\n",
    "      )\n",
    "   df_importer.node_id_property_name(\"id\")\n",
    "   df_importer.insert()\n",
    "    \n",
    "    \n",
    "print()    \n",
    "print(\"This last dataframe is large, and takes a many minutes\")    \n",
    "print(\"before you will see any output/progress ..\")    \n",
    "print()    \n",
    "print(\"Further, we are loading using Pandas dataframes from the\")    \n",
    "print(\"Jupyter container, which is the slowest choice. For speed,\")    \n",
    "print(\"we should be using Dask dataframes loaded from files\")    \n",
    "print(\"accessible from the KGIP worker nodes.\")    \n",
    "print()    \n",
    "\n",
    "    \n",
    "with import_data.DataFrameImporter(my_graph) as df_importer:   \n",
    "   df_importer.edges_dataframe(\n",
    "      pd_edges2_all,\n",
    "      source_id_space       = \"CELL_LINE\", \n",
    "      destination_id_space  = \"GENE\",   \n",
    "      source_column         = \"START_ID\",\n",
    "      destination_column    = \"END_ID\",\n",
    "      type                  = \"HAS_MUTATION_OBSERVATION\"\n",
    "      )\n",
    "\n",
    "   df_importer.node_id_property_name(\"id\")\n",
    "   df_importer.insert()\n",
    "\n",
    "print(\"\")\n",
    "print(\"--\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffb7480f-bee8-4374-b36a-806aab71e3de",
   "metadata": {},
   "source": [
    "#  Step 00: For additional diagnostics, move to the next notebook .."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05ae06a9-05a2-4feb-8ef6-f3266b96bc91",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  Check the above, runs counts\n",
    "#\n",
    "\n",
    "l_query  = \"\"\"\n",
    "   MATCH (n) \n",
    "   WITH DISTINCT LABELS(n) AS temp, COUNT(n) AS tempCnt\n",
    "   UNWIND temp AS label\n",
    "   RETURN label, SUM(tempCnt) AS cnt\n",
    "   ORDER BY label\n",
    "   \"\"\".format()\n",
    "      #\n",
    "l_result1 = my_graph.query_unpaginated(l_query)\n",
    "\n",
    "l_query  = \"\"\"\n",
    "   MATCH (m)-[r]->(n) \n",
    "   WITH DISTINCT TYPE(r) AS type, COUNT(r) AS cnt\n",
    "   RETURN type, cnt\n",
    "   ORDER BY type\n",
    "   \"\"\".format()\n",
    "      #\n",
    "l_result2 = my_graph.query_unpaginated(l_query)\n",
    "\n",
    "print()\n",
    "   #\n",
    "for l_each in l_result1.itertuples():\n",
    "   print(\"Node, %-38s   Number of records in graph: %-8d   Number of records from file: %-8d\" % (l_each.label, l_each.cnt, len(pd_nodes[l_each.label]) ))\n",
    "      #\n",
    "print()\n",
    "\n",
    "print()\n",
    "   #\n",
    "for l_each in l_result2.itertuples():\n",
    "   if (l_each.type == \"HAS_MUTATION_OBSERVATION\"):\n",
    "      print(\"Edge, %-38s   Number of records in graph: %-8d   Number of records from file: %-8d\" % (l_each.type, l_each.cnt, len(pd_edges2_all         ) ))   \n",
    "   else:\n",
    "      print(\"Edge, %-38s   Number of records in graph: %-8d   Number of records from file: %-8d\" % (l_each.type, l_each.cnt, len(pd_edges1[l_each.type]) ))   \n",
    "      #\n",
    "print()\n",
    "\n",
    "print(\"--\")\n",
    "\n",
    "\n",
    "#  Sample output,\n",
    "#\n",
    "#     Node, CELL_LINE                                Number of records in graph: 1457       Number of records from file: 1457    \n",
    "#     Node, DRUG                                     Number of records in graph: 223        Number of records from file: 223     \n",
    "#     Node, GDSC                                     Number of records in graph: 266        Number of records from file: 266     \n",
    "#     Node, GENE                                     Number of records in graph: 719        Number of records from file: 719     \n",
    "#     \n",
    "#     Edge, FOR_DRUG                                 Number of records in graph: 238        Number of records from file: 266     \n",
    "#     Edge, HAS_CELL_LINE                            Number of records in graph: 255360     Number of records from file: 257754  \n",
    "#     Edge, HAS_EXPRESSION_OBSERVATION               Number of records in graph: 391017     Number of records from file: 391017  \n",
    "#     Edge, HAS_METHYLATION_OBSERVATION              Number of records in graph: 426500     Number of records from file: 426500  \n",
    "#     Edge, HAS_MUTATION_OBSERVATION                 Number of records in graph: 12422852   Number of records from file: 12422852\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc11376b-8bca-4905-8ef3-c066f8700b52",
   "metadata": {},
   "source": [
    "#  Step 03: Create bi-directional edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6484c16-1dd3-42ac-975e-98ce0c18c360",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#  Above, we created edges in one direction.\n",
    "#\n",
    "#  Here, we use Cypher to create bi-directional edges\n",
    "#\n",
    "\n",
    "def f_getedge():\n",
    "    \n",
    "   l_query  = \"\"\"\n",
    "   \n",
    "      MATCH (n) - [r] -> (m) \n",
    "      WITH DISTINCT TYPE(r) AS edge_type, COUNT(r) AS edge_count, LABELS(n) AS src_node, LABELS(m) AS dst_node\n",
    "      RETURN edge_type, edge_count, src_node, dst_node\n",
    "      ORDER BY edge_type\n",
    "      \n",
    "      \"\"\".format()\n",
    "   \n",
    "   l_result = my_graph.query_unpaginated(l_query)\n",
    "      #\n",
    "   return l_result\n",
    "\n",
    "\n",
    "l_result = f_getedge()\n",
    "   #\n",
    "print(tabulate(l_result, headers = \"keys\", tablefmt = \"psql\", showindex = False))\n",
    "\n",
    "\n",
    "   ###\n",
    "    \n",
    "\n",
    "#  Loop thru the result set from above.\n",
    "#\n",
    "#  Two assumptions are made,\n",
    "#\n",
    "#     .  You can create the same edge multiple times, which can be bad for\n",
    "#        any analytics you run.\n",
    "#        We don't check for that here.\n",
    "#\n",
    "#     .  Vertices can have multiple lebels, hence this property coming back \n",
    "#        as an array. We process just the first label.\n",
    "#\n",
    "\n",
    "l_cntr = 0\n",
    "   #\n",
    "for l_each in l_result.itertuples():\n",
    "\n",
    "   l_cntr += 1\n",
    "      #\n",
    "   print(\"Create Edge %d of %d: %-32s   %s  -->  %s\" % (l_cntr, len(l_result), l_each.edge_type, l_each.src_node[0], l_each.dst_node[0]) )\n",
    "      #\n",
    "   l_query  = \"\"\"\n",
    "   \n",
    "      MATCH (n: {0}) - [r: {1}] -> (m: {2}) \n",
    "      WITH n, m \n",
    "      CREATE (m) - [e: {1}] -> (n)\n",
    "      \n",
    "      \"\"\".format(l_each.src_node[0], l_each.edge_type, l_each.dst_node[0])\n",
    "         #\n",
    "   l_result2 = my_graph.query_unpaginated(l_query)\n",
    "   print()\n",
    "    \n",
    "    \n",
    "l_result = f_getedge()\n",
    "   #\n",
    "print(tabulate(l_result, headers = \"keys\", tablefmt = \"psql\", showindex = False))\n",
    "\n",
    "\n",
    "print(\"--\")\n",
    "\n",
    "\n",
    "#  Sample output,\n",
    "#\n",
    "#     +-----------------------------+--------------+---------------+---------------+\n",
    "#     | edge_type                   |   edge_count | src_node      | dst_node      |\n",
    "#     |-----------------------------+--------------+---------------+---------------|\n",
    "#     | FOR_DRUG                    |          238 | ['GDSC']      | ['DRUG']      |\n",
    "#     | HAS_CELL_LINE               |       255360 | ['GDSC']      | ['CELL_LINE'] |\n",
    "#     | HAS_EXPRESSION_OBSERVATION  |       391017 | ['CELL_LINE'] | ['GENE']      |\n",
    "#     | HAS_METHYLATION_OBSERVATION |       426500 | ['CELL_LINE'] | ['GENE']      |\n",
    "#     | HAS_MUTATION_OBSERVATION    |     12422852 | ['CELL_LINE'] | ['GENE']      |\n",
    "#     +-----------------------------+--------------+---------------+---------------+\n",
    "#     \n",
    "#     Create Edge 1 of 5: FOR_DRUG                           GDSC  -->  DRUG\n",
    "#     Create Edge 2 of 5: HAS_CELL_LINE                      GDSC  -->  CELL_LINE\n",
    "#     Create Edge 3 of 5: HAS_EXPRESSION_OBSERVATION         CELL_LINE  -->  GENE\n",
    "#     Create Edge 4 of 5: HAS_METHYLATION_OBSERVATION        CELL_LINE  -->  GENE\n",
    "#     Create Edge 5 of 5: HAS_MUTATION_OBSERVATION           CELL_LINE  -->  GENE\n",
    "#     \n",
    "#     +-----------------------------+--------------+---------------+---------------+\n",
    "#     | edge_type                   |   edge_count | src_node      | dst_node      |\n",
    "#     |-----------------------------+--------------+---------------+---------------|\n",
    "#     | FOR_DRUG                    |          238 | []            | []            |\n",
    "#     | FOR_DRUG                    |          238 | ['GDSC']      | ['DRUG']      |\n",
    "#     | HAS_CELL_LINE               |       255360 | ['GDSC']      | ['CELL_LINE'] |\n",
    "#     | HAS_CELL_LINE               |       255360 | []            | []            |\n",
    "#     | HAS_EXPRESSION_OBSERVATION  |       391017 | []            | []            |\n",
    "#     | HAS_EXPRESSION_OBSERVATION  |       391017 | ['CELL_LINE'] | ['GENE']      |\n",
    "#     | HAS_METHYLATION_OBSERVATION |       426500 | ['CELL_LINE'] | ['GENE']      |\n",
    "#     | HAS_METHYLATION_OBSERVATION |       426500 | []            | []            |\n",
    "#     | HAS_MUTATION_OBSERVATION    |     12422852 | []            | []            |\n",
    "#     | HAS_MUTATION_OBSERVATION    |     12422852 | ['CELL_LINE'] | ['GENE']      |\n",
    "#     +-----------------------------+--------------+---------------+---------------+\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae597fe0-50fa-4fc3-98e6-44e5abbfcc7f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4d6ddf3-a46c-4461-adff-ffa95b586895",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
